core/src/test/java/org/apache/iceberg/TestSequenceNumberForV2Table.java

```diff
@@ -0,0 +1,328 @@
+    txn.commitTransaction();
+
+    Assert.assertEquals("Snapshot sequence number should match expected",
+        2, table.currentSnapshot().sequenceNumber());
+    validateDataFiles(files(FILE_B), seqs(2));
+  }
+
+  @Test
+  public void testTransactionFailure() {
+    table.newAppend()
+        .appendFile(FILE_A)
+        .appendFile(FILE_B)
+        .commit();
+
+    table.updateProperties()
+        .set(TableProperties.COMMIT_NUM_RETRIES, "0")
+        .commit();
+
+    table.ops().failCommits(1);
+
+    Transaction txn = table.newTransaction();
+    txn.newAppend().appendFile(FILE_C).commit();
+
+    AssertHelpers.assertThrows("Transaction commit should fail",
+        CommitFailedException.class, "Injected failure", txn::commitTransaction);
+
+    Assert.assertEquals("Snapshot sequence number should match expected",
+        1, table.currentSnapshot().sequenceNumber());
+  }
+
+  @Test
+  public void testCherryPicking() {
+    table.newAppend()
+        .appendFile(FILE_A)
+        .commit();
+
+    table.newAppend()
+        .appendFile(FILE_B)
+        .stageOnly()
+        .commit();
+
+    Assert.assertEquals("Snapshot sequence number should be 1", 1,
+        table.currentSnapshot().sequenceNumber());
+
+    // pick the snapshot that's staged but not committed
+    Snapshot stagedSnapshot = readMetadata().snapshots().get(1);
+
+    Assert.assertEquals("Snapshot sequence number should be 2", 2,
+        stagedSnapshot.sequenceNumber());
+
+    // table has new commit
+    table.newAppend()
+        .appendFile(FILE_C)
+        .commit();
+
+    Assert.assertEquals("Snapshot sequence number should be 3",
+        3, table.currentSnapshot().sequenceNumber());
+
+    // cherry-pick snapshot
+    table.manageSnapshots().cherrypick(stagedSnapshot.snapshotId()).commit();
+
+    Assert.assertEquals("Snapshot sequence number should be 4",
+        4, table.currentSnapshot().sequenceNumber());
+
+
+    validateDataFiles(files(FILE_A, FILE_B, FILE_C), seqs(1, 4, 3));
+  }
+
+  @Test
+  public void testCherryPickFastForward() {
+    table.newAppend()
+        .appendFile(FILE_A)
+        .commit();
+
+    table.newAppend()
+        .appendFile(FILE_B)
+        .stageOnly()
+        .commit();
+
+    Assert.assertEquals("Snapshot sequence number should be 1", 1,
+        table.currentSnapshot().sequenceNumber());
+
+    // pick the snapshot that's staged but not committed
+    Snapshot stagedSnapshot = readMetadata().snapshots().get(1);
+
+    Assert.assertEquals("Snapshot sequence number should be 2", 2,
+        stagedSnapshot.sequenceNumber());
+
+    // cherry-pick snapshot, this will fast forward
+    table.manageSnapshots().cherrypick(stagedSnapshot.snapshotId()).commit();
+    Assert.assertEquals("Snapshot sequence number should be 2",
+        2, table.currentSnapshot().sequenceNumber());
+
+    validateDataFiles(files(FILE_A, FILE_B), seqs(1, 2));
+  }
+
+  void validateDataFiles(Iterator<DataFile> files, Iterator<Long> expectedSeqs) {
```

Comment:
These tests should use the validation methods from TableTestBase instead of adding its own. This method causes the test cases in this PR to be confusing because sometimes the sequence number is for a delete. Instead, the test cases need to be very specific about metadata.

All the metadata for every change should be validated:
* Each operation that produces a sequence number should assert the number of changed manifests
* Each changed manifest should have its entries validated, including file status, snapshot id, and sequence number
* Operations that do not produce a sequence number should validate that the sequence number did not change using the last sequence number in table metadata.

Reply:
OK, will update accordingly.